src/main/java/at/ac/tuwien/kr/alpha/solver/learning/GroundConflictNoGoodLearner.java

```diff
@@ -110,174 +117,144 @@ public GroundConflictNoGoodLearner(Assignment assignment) {
-		// Resolve implied nogood into current resolution.
-		int resolvedLiterals[] = new int[secondNoGood.size() + firstNoGood.size() - 2];
-		int resolvedCounter = 0;
-		// Copy over all literals except the resolving ones.
-		for (int i = 0; i < firstNoGood.size(); i++) {
-			if (atomOf(firstNoGood.getLiteral(i)) != resolutionAtom) {
-				resolvedLiterals[resolvedCounter++] = firstNoGood.getLiteral(i);
+			// Due to out-of-order assigned literals, the learned nogood may be not assigning.
+			backjumpingDecisionLevel = computeConflictFreeBackjumpingLevel(learnedNoGood);
+			if (backjumpingDecisionLevel < 0) {
+				return ConflictAnalysisResult.UNSAT;
 			}
 		}
-		// Copy literals from implying nogood except the resolving one and sort additional literals into processing list.
-		for (int i = 0; i < secondNoGood.size(); i++) {
-			if (atomOf(secondNoGood.getLiteral(i)) != resolutionAtom) {
-				resolvedLiterals[resolvedCounter++] = secondNoGood.getLiteral(i);
+		LOGGER.trace("Backjumping decision level: {}", backjumpingDecisionLevel);
+		return new ConflictAnalysisResult(learnedNoGood, backjumpingDecisionLevel, resolutionAtoms, computeLBD(learnedLiterals));
+	}
 
-				// Sort literal also into queue for further processing.
-				Assignment.Entry newLiteral = getAssignmentEntryRespectingLowerMBT(secondNoGood.getLiteral(i));
-				if (assignment instanceof TrailAssignment) {
-					firstUIPPriorityQueue.add(newLiteral);
-					continue;
-				}
-				// Check for special case where literal was assigned from MBT to TRUE on the same decisionLevel and the propagationLevel of TRUE is higher than the one of the resolutionLiteral.
-				if (newLiteral.getDecisionLevel() == resolutionLiteralDecisionLevel
-					&& newLiteral.getPropagationLevel() > resolutionLiteralPropagationLevel) {
-					if (newLiteral.hasPreviousMBT()
-						&& newLiteral.getMBTDecisionLevel() == resolutionLiteralDecisionLevel
-						&& newLiteral.getMBTPropagationLevel() < resolutionLiteralPropagationLevel) {
-						// Resort to the previous entry (i.e., the one for MBT) and use that one.
-						firstUIPPriorityQueue.add(newLiteral);
-						continue;
-					}
-					throw oops("Implying literal on current decisionLevel has higher propagationLevel than the implied literal and this was no assignment from MBT to TRUE");
-				}
-				// Add literal to queue for finding 1UIP.
-				firstUIPPriorityQueue.add(newLiteral);
+	private int computeLBD(int[] literals) {
+		HashSet<Integer> occurringDecisionLevels = new HashSet<>();
+		for (int literal : literals) {
+			if (!assignment.isAssigned(atomOf(literal))) {
+				throw oops("Atom is not assigned: " + atomOf(literal));
 			}
+			occurringDecisionLevels.add(assignment.getWeakDecisionLevel(atomOf(literal)));
 		}
-		return resolvedLiterals;
+		return occurringDecisionLevels.size();
 	}
 
 	public ResolutionSequence obtainResolutionSequence() {
```

Comment:
This method is never called, we could remove it

Reply:
Removed.